clients/src/main/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadata.java

```diff
@@ -58,25 +59,55 @@ public boolean isCancelled() {
     @Override
     public RecordMetadata get() throws InterruptedException, ExecutionException {
         this.result.await();
+        if (nextRecordMetadata != null)
+            return nextRecordMetadata.get();
         return valueOrError();
     }
 
     @Override
     public RecordMetadata get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
+        // Take the max to handle overflow.
```

Comment:
Sounds good. Minor: may want to switch to `Long.MAX_VALUE` on overflow for `deadline`.